cli/src/org/partiql/cli/functions/ReadFile.kt

```diff
@@ -88,6 +95,92 @@ internal class ReadFile(valueFactory: ExprValueFactory) : BaseFunction(valueFact
+            get() = lazyIonValue
+
+        /** Laziness is even more important here because [ionValue] is likely to never be called. */
+        private val lazyIonValue by lazy {
+            valueFactory.ion.newEmptyStruct().apply {
+                rowValues.map { kvp ->
+                    if (kvp.value.scalar.numberValue() != null) {
+                        add(kvp.key, valueFactory.ion.newInt(kvp.value.scalar.numberValue()))
+                    } else {
+                        add(kvp.key, valueFactory.ion.newString(kvp.value.scalar.stringValue()))
+                    }
+                }
+                makeReadOnly()
+            }
+        }
+
+        /**
+         * The lazily constructed set of row values.  Laziness is good practice here because it avoids
+         * constructing this map if it isn't needed.
+         */
+        private val rowValues: Map<String, ExprValue> by lazy {
+            rowString.trim().drop(1).dropLast(1)
+                .split('|')
+                .mapIndexed { i, it ->
+                    val fieldName = header[i]
+
+                    val value = it.trim()
+                    // Note that we invoke
+                    if (value[0] == '\"') {
+                        fieldName to valueFactory.newString(value.drop(1).dropLast(1)).namedValue(valueFactory.newString(fieldName))
+                    } else {
+                        fieldName to valueFactory.newInt(value.toInt()).namedValue(valueFactory.newString(fieldName))
+                    }
+                }.toMap()
+        }
+
+        /** An iterator over the values contained in this instance of [ExprValue], if any. */
+        override fun iterator() = rowValues.values.iterator()
+
+        private val bindingsInstance by lazy {
+            Bindings.ofMap(rowValues)
+        }
+
+        override val bindings: Bindings<ExprValue>
+            get() = bindingsInstance
+    }
+    /**
+     * The current supported format for readFile function is
+     * read_file('simple.csv', {'type':'csv'});
+     * Extend the support to markdown
+     */
+    private fun mdReadHandler(): (InputStream, IonStruct) -> ExprValue = { input, options ->
+        val reader = InputStreamReader(input)
+        val ion = IonSystemBuilder.standard().build()
+        val valueFactory = ExprValueFactory.standard(ion)
+        val pipeline = CompilerPipeline.standard(valueFactory)
+        // get header
+        val rowData = reader.readLines()
+        val header = rowData[0]
+            .trim()
+            .drop(1)
+            .dropLast(1)
+            .split('|')
+            .map { it ->
+                it.trim()
+            }
+
+        // a bag is a sequence of ExprValue, so create a Mdrowexprvalue class
+        valueFactory.newBag(
+            rowData.asSequence()
+                .filter { it -> it.isNotEmpty() }
+                .drop(2)
+                .map { it ->
+                    MDRowExprValue(pipeline.valueFactory, it, header)
```

Comment:
Can we directly construct a struct, instead of creating a new class `MDRowExprValue`? You can refer [here](https://github.com/partiql/partiql-lang-kotlin/blob/main/lang/src/org/partiql/lang/eval/io/DelimitedValues.kt#L91-#L102) to see how to directly construct an `ExprValue` with struct type.